File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        int left = 0;
        long long ans = 0;
        long long curr = 0;

        unordered_set<int> hashset;

        for (int right =0; right < nums.size(); right++){

            curr += nums[right];

            while (hashset.count(nums[right])){
                hashset.erase(nums[left]);
                curr -= nums[left];
                left += 1;
            }

            hashset.insert(nums[right]);

            if (right-left+1==k){
                ans = max(ans, curr);
                curr -= nums[left];
                hashset.erase(nums[left]);
                left += 1;
            }

        }
        return ans;
        
    }
};

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.py
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:

        hashSet = set()

        left = 0
        ans = 0
        curr = 0
        for i, num in enumerate(nums):
            if num in hashSet:
                while num in hashSet:
                    hashSet.remove(nums[left])
                    curr -= nums[left]
                    left += 1
            curr += num
            hashSet.add(num)
            if i-left+1 == k:
                ans = max(ans, curr)
                hashSet.remove(nums[left])
                curr -= nums[left]
                left += 1
        return ans
        